src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/ExactMethodInstantiationsNode.cs

```diff
@@ -0,0 +1,173 @@
+                return new ObjectData(Array.Empty<byte>(), Array.Empty<Relocation>(), 1, new ISymbolNode[] { this });
+
+            // Zero out the hashset so that we AV if someone tries to insert after we're done.
+            _exactMethodInstantiationsList = null;
+
+            // Ensure the native layout data has been saved, in order to get valid Vertex offsets for the signature Vertices
+            factory.MetadataManager.NativeLayoutInfo.SaveNativeLayoutInfoWriter(factory);
+
+            MemoryStream stream = new MemoryStream();
+            _nativeWriter.Save(stream);
+
+            byte[] streamBytes = stream.ToArray();
+
+            _endSymbol.SetSymbolOffset(streamBytes.Length);
+
+            return new ObjectData(streamBytes, Array.Empty<Relocation>(), 1, new ISymbolNode[] { this, _endSymbol });
+        }
+
+        public static DependencyList GetExactMethodInstantiationDependenciesForMethod(NodeFactory factory, MethodDesc method)
+        {
+            if (!IsMethodEligibleForTracking(method))
+                return null;
+
+            DependencyList dependencies = new DependencyList();
+
+            // Method entry point dependency
+            bool getUnboxingStub = method.OwningType.IsValueType && !method.Signature.IsStatic;
+            IMethodNode methodEntryPointNode = factory.MethodEntrypoint(method, getUnboxingStub);
+            dependencies.Add(new DependencyListEntry(methodEntryPointNode, "Exact method instantiation entry"));
+
+            // Get native layout dependencies for the declaring type
+            dependencies.Add(new DependencyListEntry(factory.NecessaryTypeSymbol(method.OwningType), "Exact method instantiation entry"));
+
+            // Get native layout dependencies for the method instantiation args
+            foreach (var arg in method.Instantiation)
+                dependencies.Add(new DependencyListEntry(factory.NecessaryTypeSymbol(arg), "Exact method instantiation entry"));
+
+            // Get native layout dependencies for the method signature.
+            NativeLayoutMethodNameAndSignatureVertexNode nameAndSig = factory.NativeLayoutMethodNameAndSignatureVertex(method.GetTypicalMethodDefinition());
+            dependencies.Add(new DependencyListEntry(factory.NativeLayoutPlacedSignatureVertex(nameAndSig), "Exact method instantiation entry"));
+
+            return dependencies;
+        }
+
+        public void AddEntryIfEligible(NodeFactory factory, MethodDesc method)
+        {
+            // Check if we already saw this method
+            if (!_exactMethodInstantiationsList.Add(method))
+                return;
+
+            if (!IsMethodEligibleForTracking(method))
+                return;
+
+            bool getUnboxingStub = method.OwningType.IsValueType && !method.Signature.IsStatic;
+            IMethodNode methodEntryPointNode = factory.MethodEntrypoint(method, getUnboxingStub);
+
+            Vertex methodSignature = GetNativeLayoutSignatureForMethodEntry(factory, method);
```

Comment:
We're getting the Vertex too early here. `AddEntryIfEligible` is called as we're compiling things - this will have the problem with stability of indices in the ExternalReferencesTable in multithreaded compilation that we discussed yesterday.

Should entries in the hash table be on the same plan as the other NativeLayoutVertexNodes?